lib/dialects/abstract/query-generator.js

```diff
@@ -723,8 +724,17 @@ module.exports = (function() {
             } else {
               joinOn = self.quoteIdentifier(tableLeft) + "." + self.quoteIdentifier(attrLeft) + " = "
             }
-            joinOn += self.quoteIdentifier(tableRight) + "." + self.quoteIdentifier(attrRight)
-
+            joinOn += fullAttrRight
+
+            if (
+              subQuery &&
+              include.parent.table === self.quoteIdentifier(parentTable) &&
+              tableRight === parentTable &&
+              subQueryAttributes.indexOf(fullAttrRight) < 0 &&
+              subQueryAttributes.indexOf(self.quoteIdentifier(tableRight) + '.*') < 0
+              ) {
+              subQueryAttributes = subQueryAttributes.concat([fullAttrRight]);
+            }
```

Comment:
I have to confess that I arrived at this using a shameful amount of trial and error.